pkg/component/networking/nodelocaldns/nodelocaldns.go

```diff
@@ -158,368 +190,181 @@ func (n *nodeLocalDNS) Deploy(ctx context.Context) error {
 		return err
 	}
 
-	data, err := n.computeResourcesData()
+	serviceAccount, configMap, service, err := n.computeResourcesData()
+	if err != nil {
+		return err
+	}
+
+	data, err := registry.AddAllAndSerialize(n.computePoolResourcesData(serviceAccount, configMap, service)...)
 	if err != nil {
 		return err
 	}
+
 	return managedresources.CreateForShoot(ctx, n.client, n.namespace, managedResourceName, managedresources.LabelValueGardener, false, data)
 }
 
 func (n *nodeLocalDNS) Destroy(ctx context.Context) error {
+	managedResourceExists := false
+	// Check if the managed resource exists
+	if err := n.client.Get(ctx, types.NamespacedName{Namespace: n.namespace, Name: managedResourceName}, &resourcesv1alpha1.ManagedResource{}); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return fmt.Errorf("failed to check managed resource existence: %w", err)
+		}
+	} else {
+		managedResourceExists = true
+	}
+
+	// Check if at least one Kubernetes node has the cleanup label
+	nodeList := &corev1.NodeList{}
+	if err := n.values.ShootClient.List(ctx, nodeList, client.MatchingLabels{
+		labelKeyCleanupRequired: "true",
+	}); err != nil {
+		return fmt.Errorf("failed to list nodes with cleanup label: %w", err)
+	}
+
+	// If the managed resource does not exist and no nodes have the cleanup label, return early no cleanup needed
+	if !managedResourceExists && len(nodeList.Items) == 0 {
+		return nil
+	}
+
+	// Mark nodes for cleanup if the managed resource exists
+	if managedResourceExists {
+		if err := n.markNodeForCleanup(ctx); err != nil {
+			return fmt.Errorf("failed to mark nodes for cleanup: %w", err)
+		}
+	}
+
+	// Delete resources
 	if err := kubernetesutils.DeleteObjects(ctx, n.client,
 		n.emptyScrapeConfig(""),
 		n.emptyScrapeConfig("-errors"),
 	); err != nil {
 		return err
 	}
 
-	return managedresources.DeleteForShoot(ctx, n.client, n.namespace, managedResourceName)
-}
+	if err := managedresources.DeleteForShoot(ctx, n.client, n.namespace, managedResourceName); err != nil {
+		return err
+	}
 
-// TimeoutWaitForManagedResource is the timeout used while waiting for the ManagedResources to become healthy
-// or deleted.
-var TimeoutWaitForManagedResource = 2 * time.Minute
+	if v1beta1helper.IsKubeProxyIPVSMode(n.values.KubeProxyConfig) {
+		return nil
+	}
 
-func (n *nodeLocalDNS) Wait(ctx context.Context) error {
+	// Wait until the managed resource is successfully deleted and go through node list again to add cleanup label
 	timeoutCtx, cancel := context.WithTimeout(ctx, TimeoutWaitForManagedResource)
 	defer cancel()
+	if err := managedresources.WaitUntilDeleted(timeoutCtx, n.client, n.namespace, managedResourceName); err != nil {
+		return err
+	}
 
-	return managedresources.WaitUntilHealthy(timeoutCtx, n.client, n.namespace, managedResourceName)
-}
+	nodeList = &corev1.NodeList{}
+	if err := n.values.ShootClient.List(ctx, nodeList); err != nil {
```

Comment:
You could use a label selector to only list nodes that don't have the label yet. This way, you can remove L268-271 entirely.